### ISSUE TITLE
Asynchronously wait for all workers to finish

### ISSUE THREAD

**Issue by SUPERCILEX** [gradle/gradle#9377]
AFAIK, the only wait to wait for all workers to finish is to block the task action. I'd like a way to submit another worker that only executes when all the others are done. Currently, I have to either use a single worker for my task that lets other tasks do their thing, or I use a bunch of workers but I have to block my task which prevents other tasks from doing their things.

**Comment by SUPERCILEX** [comment]
@ghale is this possible?

**Comment by ghale** [comment]
So, the issue here is that once you release other tasks to run, the project state can change.  We made the decision that we did not want to release the project lock _inside_ of a task action because we wanted to allow you to make the assumption that nothing would change the project state while your task action is running except your task itself.  On the other hand, if you wanted to allow some state change between one region of logic and another, you could simply split it into multiple task actions (ie, the first runs, submits work and finishes, other tasks can run while the async work is running, and then once the async work finishes, the next task action runs and can submit more work).  

Having said that, the decision to not release the project lock inside of a task action was a bit of a judgement call and I'm not sure that we have a strong sense of the use cases around this.  For instance, when would it be necessary to have a single task action rather than just splitting the work up into multiple task actions?  It would be interesting to know a little more about your use case to better understand how people are using this capability.

**Comment by SUPERCILEX** [comment]
First off, thanks for the thorough explanation! And I'm definitely not advocating against the project lock (I think things would be a 💩 show if extensions and their values could change from outside while running a task action).

What I'm asking for in a more general sense is the ability to force a deterministic worker execution order. I don't think it makes sense to be able to order each individual worker, but being able to say _this worker should run before/after these other ones_ would be nice. To give a little more details on my use case, I'm starting my task action by generating an ID for every worker to share, then kicking off the workers, then waiting for all workers to finish such that I can commit the changes each worker applied to the ID. This shouldn't have to block the task since I can just give the last worker to be executed the ID and it can figure out how to commit the changes as long as it knows the other workers are done.

TL;DR: more control over worker execution order.

**Comment by SUPERCILEX** [comment]
I actually thought of a really nice trick: https://github.com/Triple-T/gradle-play-publisher/commit/7a8f57a1a94a098b6537962c4746bcd05f8467d8
